content/articles/gulp4-primer.md

```diff
@@ -0,0 +1,131 @@
+To solve that very common issue of task dependencies, 2 methods were added:
+
+- gulp.series: will run the tasks in order
+- gulp.parallel: will run the tasks in parallel
+
+Let's fix our error by replacing those dependent tasks, for example for the default task:
+
+```js
+gulp.task(
+  'default',
+  gulp.series('build', gulp.parallel('browser-sync', 'watch', 'karma-watch'))
+);
+```
+
+In that case we first want to run build, and then browser-sync, watch and karma-watch in parallel. Easy (the commit below will contain a buggy version since I did put browser-sync in the series by mistake, *easy*)
+Now trying to run the build task, I get another error:
+```js
+[15:09:07] TypeError: undefined is not a function
+    at /home/vincent/Code/ng-boilerplate/node_modules/run-sequence/index.js:18:22
+    at Array.forEach (native)
+    at verifyTaskSets (/home/vincent/Code/ng-boilerplate/node_modules/run-sequence/index.js:12:11)
+```
+Since gulp 4 has ways to run things sequentially, let's use that.
+```js
+// before
+gulp.task('build', function(){
+  return runSequence(
+    'clean',
+    ['sass', 'copy-assets', 'ts-compile', 'templates', 'copy-vendor'],
+    'index'
+  );
+)};
+// after
+gulp.task(
+  'build',
+  gulp.series(
+    'clean',
+    gulp.parallel('sass', 'copy-assets', 'ts-compile', 'templates', 'copy-vendor'),
+    'index'
+  )
+);
+```
+
+Note: I had to modify some tasks like the browser-sync since its API changed and had to make a [patch](https://github.com/dlmanning/gulp-sass/pull/207) to gulp-sass to update the node-sass version.
+You can see the commit [here](https://github.com/Keats/ng-boilerplate/commit/91a05401c6ea532467bc00b5a6e54fd95b6b0eaf).
+Looking at the [build](https://travis-ci.org/Keats/ng-boilerplate/builds/54274589), it still fails because of a missing selenium jar so a protractor error.
+My elegant solution was the [following](https://github.com/Keats/ng-boilerplate/commit/d999db2442ac72fc440e8ff9cccd8348c533c72d), as in removing protractor from travis because I'm not using it at point anyway.
+
+### Small cleanup
+This part is not strictly related to gulp4 but contains some of the things I like to do to have a clean gulpfile.
+The config part at the top is quite ugly but I am not going to touch it here, only change the gulp related things.
+
+First thing, [gulp-load-plugins](https://www.npmjs.com/package/gulp-load-plugins) makes things quite simple but the prefix I had chosen, 'plugins', is quite verbose and takes up space. Recently I've started using $ which works quite well.
+Another point that took me way too long to realise it auto-camelCase packages names with dash so my following line was useless
+```js
+plugins.ngAnnotate = require('gulp-ng-annotate');
+```
+Another thing is to look closely to whether some of the gulp-plugins you're using actually bring any values compared to using the actual thing directly.
```

Comment:
Another thing to examine is whether your gulp-plugins still provide value over using the ..

And what is "the thing" in this context? Not specific enough.